README.md
# YoutubeSourceParserKit 

==================
[![Build Status](https://img.shields.io/travis/movielala/YoutubeSourceParserKit/master.svg)](https://travis-ci.org/movielala/YoutubeSourceParserKit)
![License](https://img.shields.io/badge/license-MIT-blue.svg)
[![Dependencies](https://img.shields.io/badge/dependencies-none-brightgreen.svg)](https://github.com/mobileplayer/mobileplayer-ios)
[![Ready](https://badge.waffle.io/movielala/YoutubeSourceParserKit.png?label=Ready&title=Ready)](https://waffle.io/movielala/YoutubeSourceParserKit)
[![StackOverflow](https://img.shields.io/badge/StackOverflow-Ask%20a%20question!-blue.svg)](http://stackoverflow.com/questions/ask?tags=YoutubeSourceParserKit+ios+swift)
[![CocoaPods](https://img.shields.io/cocoapods/v/YoutubeSourceParserKit.svg)](https://img.shields.io/cocoapods/v/YoutubeSourceParserKit.svg)
[![Join the chat at https://gitter.im/mobileplayer/mobileplayer-ios](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/movielala/YoutubeSourceParserKit)


YouTube Video Link Parser for Swift. Heavily inspried from hellozimi's repo [HCYoutubeParser](https://github.com/hellozimi/HCYoutubeParser)

##Introduction

__Requires iOS 8 or later and Xcode 7.0+__<br/>
Swift support uses dynamic frameworks and is therefore only supported on iOS > 8. Version 0.2.5 adds swift 3.0 support.

##Installation

To install via CocoaPods add this line to your `Podfile`.
```
use_frameworks!
```
and
```
pod 'YoutubeSourceParserKit'
```

Then, run the following command:

```$ pod install```

##Usage

```swift
import YoutubeSourceParserKit
```

```swift
let testURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")!
    Youtube.h264videosWithYoutubeURL(testURL) { (videoInfo, error) -> Void in
      if let videoURLString = videoInfo?["url"] as? String,
        videoTitle = videoInfo?["title"] as? String {
          print("\(videoTitle)")
          print("\(videoURLString)")
      }
    }
```

```
videoInfo output:
```
```json
{
    "title": "[Video Title]",
    "isStream": 0,
    "quality": "hd720",
    "itag": 22,
    "fallback_host": "tc.v20.cache2.googlevideo.com",
    "url": "http://[Source URL]"
}
```

##MPMoviePlayerController Usage

![alt tag](http://s10.postimg.org/5j1mristl/i_OS_Simulator_Screen_Shot_Jul_12_2015_14_33_02.png)

```swift
import UIKit
import YoutubeSourceParserKit
import MediaPlayer

class ViewController: UIViewController {

  let moviePlayer = MPMoviePlayerController()

  override func viewDidLoad() {
    super.viewDidLoad()
    moviePlayer.view.frame = view.frame
    view.addSubview(moviePlayer.view)
    moviePlayer.fullscreen = true
    let youtubeURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")!
    playVideoWithYoutubeURL(youtubeURL)
  }

  func playVideoWithYoutubeURL(url: NSURL) {
    Youtube.h264videosWithYoutubeURL(url, completion: { (videoInfo, error) -> Void in
      if let
        videoURLString = videoInfo?["url"] as? String,
        videoTitle = videoInfo?["title"] as? String {
          self.moviePlayer.contentURL = NSURL(string: videoURLString)
      }
    })
  }
}
```
Warning:
This repo is for educational purposes.  This is not approved by the ToC of YouTube. Use at own risk. 


CHANGELOG.md
#### Version 0.2.5

* Support for Swift 3.0

#### Version 0.2.4

* Rename 'YoutubeSourceParserKit'
* Add tvOS support

#### Version 0.2.3

* Supporting Swift 2.0

#### Version 0.2.2

* Add Changelog file

#### Version 0.2.1

* Fixed gcd bug

#### Version 0.2.0

* Add gcd to parser operations

#### Version 0.1.9

* Updated project setting and podspec file

#### Version 0.1.8

* Supporting for iOS7

#### Version 0.1.7

* Add test conditions and update podspec file

#### Version 0.1.6

* Update README.md

#### Version 0.1.5

* Update podspec file

#### Version 0.1.4

* Update podspec file

#### Version 0.1.3

* Add MPMoviePlayerController Usage

#### Version 0.1.2

* Update podspec file

#### Version 0.1.1

* Add youtube-parser class

#### Version 0.1.0

* Initial version


AppDelegate.swift
//
//  AppDelegate.swift
//  youtube-parser
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit
import Photos

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {
  
    var window: UIWindow?
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey : Any]? = nil) -> Bool {
        //let testURL = NSURL(string: "https://www.youtube.com/watch?v=swZJwZeMesk")! // copyrighted content
        let testURL = NSURL(string: "https://www.youtube.com/watch?v=MmyGP03tt0c")! // free content
        Youtube.h264videosWithYoutubeURL(youtubeURL: testURL) { (videoInfo, error) -> Void in
            if let videoURLString = videoInfo?["url"] as? String,
                let videoTitle = videoInfo?["title"] as? String {
                print("\(videoTitle)")
                print("\(videoURLString)")
                
                // Save the video
                self.downloadVideoLinkAndCreateAsset(videoURLString)
            }
        }
        return true
    }
    
    func downloadVideoLinkAndCreateAsset(_ videoLink: String) {
        
        // use guard to make sure you have a valid url
        guard let videoURL = URL(string: videoLink) else { return }
        
        guard let documentsDirectoryURL = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else { return }
        
        // check if the file already exist at the destination folder if you don't want to download it twice
        if !FileManager.default.fileExists(atPath: documentsDirectoryURL.appendingPathComponent(videoURL.lastPathComponent).path) {
            
            // set up your download task
            URLSession.shared.downloadTask(with: videoURL) { (location, response, error) -> Void in
                
                // use guard to unwrap your optional url
                guard let location = location else { return }
                
                // create a deatination url with the server response suggested file name
                let destinationURL = documentsDirectoryURL.appendingPathComponent(response?.suggestedFilename ?? videoURL.lastPathComponent)
                
                do {
                    
                    try FileManager.default.moveItem(at: location, to: destinationURL)
                    
                    PHPhotoLibrary.requestAuthorization({ (authorizationStatus: PHAuthorizationStatus) -> Void in
                        
                        // check if user authorized access photos for your app
                        if authorizationStatus == .authorized {
                            PHPhotoLibrary.shared().performChanges({
                                PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: destinationURL)}) { completed, error in
                                    if completed {
                                        print("Video asset created")
                                    } else {
                                        print(error.debugDescription)
                                    }
                            }
                        }
                    })
                    
                } catch { print(error) }
                
                }.resume()
            
        } else {
            print("File already exists at destination url")
        }
        
    }
}


YoutubeSourceParserKit/Youtube.swift
//
//  Youtube.swift
//  youtube-parser
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit

public extension NSURL {
  /**
  Parses a query string of an NSURL

  @return key value dictionary with each parameter as an array
  */
  func dictionaryForQueryString() -> [String: AnyObject]? {
    if let query = self.query {
      return query.dictionaryFromQueryStringComponents()
    }

    // Note: find youtube ID in m.youtube.com "https://m.youtube.com/#/watch?v=1hZ98an9wjo"
    let result = (absoluteString! as NSString).components(separatedBy: "?")
    if result.count > 1 {
      return result.last?.dictionaryFromQueryStringComponents()
    }
    return nil
  }
}

public extension NSString {
  /**
  Convenient method for decoding a html encoded string
  */
  func stringByDecodingURLFormat() -> String {
    let result = self.replacingOccurrences(of: "+", with:" ") as NSString
    return result.removingPercentEncoding!
  }

  /**
  Parses a query string

  @return key value dictionary with each parameter as an array
  */
  func dictionaryFromQueryStringComponents() -> [String: AnyObject] {
    var parameters = [String: AnyObject]()
    for keyValue in components(separatedBy: "&") {
        let keyValueArray = keyValue.components(separatedBy: "=")
      if keyValueArray.count < 2 {
        continue
      }
      let key = keyValueArray[0].stringByDecodingURLFormat()
      let value = keyValueArray[1].stringByDecodingURLFormat()
        parameters[key] = value as AnyObject
    }
    return parameters
  }
}

public class Youtube: NSObject {
  static let infoURL = "http://www.youtube.com/get_video_info?video_id="
  static var userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4"
  /**
  Method for retrieving the youtube ID from a youtube URL

  @param youtubeURL the the complete youtube video url, either youtu.be or youtube.com
  @return string with desired youtube id
  */
  public static func youtubeIDFromYoutubeURL(youtubeURL: NSURL) -> String? {
    if let
      youtubeHost = youtubeURL.host,
        let youtubePathComponents = youtubeURL.pathComponents {
        let youtubeAbsoluteString = youtubeURL.absoluteString! as NSString
        if youtubeHost == "youtu.be" as String? {
          return youtubePathComponents[1]
        } else if (youtubeAbsoluteString as String).range(of: "www.youtube.com/embed") != nil {
          return youtubePathComponents[2]
        } else if youtubeHost == "youtube.googleapis.com" ||
          youtubeURL.pathComponents!.first == "www.youtube.com" as String? {
            return youtubePathComponents[2]
        } else if let
          queryString = youtubeURL.dictionaryForQueryString(),
            let searchParam = queryString["v"] as? String {
            return searchParam
        }
    }
    return nil
  }
  /**
  Method for retreiving a iOS supported video link

  @param youtubeURL the the complete youtube video url
  @return dictionary with the available formats for the selected video
  
  */
  public static func h264videosWithYoutubeID(youtubeID: String) -> [String: AnyObject]? {
    let urlString = String(format: "%@%@", infoURL, youtubeID) as String
    let url = NSURL(string: urlString)!
    let request = NSMutableURLRequest(url: url as URL)
    request.timeoutInterval = 5.0
    request.setValue(userAgent, forHTTPHeaderField: "User-Agent")
    request.httpMethod = "GET"
    var responseString = NSString()
    let session = URLSession(configuration: URLSessionConfiguration.default)
    let group = DispatchGroup()
    group.enter()
    session.dataTask(with: request as URLRequest, completionHandler: { (data, response, error) -> Void in
        if error != nil {
            print(error.debugDescription)
        }
        if data != nil {
            responseString = NSString(data: data!, encoding: String.Encoding.utf8.rawValue)!
        }
        group.leave()
    }).resume()
    let _ = group.wait(timeout: .distantFuture)
    //dispatch_group_wait(group, dispatch_time_t(DispatchTime.distantFuture))
    let parts = responseString.dictionaryFromQueryStringComponents()
    if parts.count > 0 {
      var videoTitle: String = ""
      var streamImage: String = ""
      if let title = parts["title"] as? String {
        videoTitle = title
      }
      if let image = parts["iurl"] as? String {
        streamImage = image
      }
      if let fmtStreamMap = parts["url_encoded_fmt_stream_map"] as? String {
        // Live Stream
        if let _: AnyObject = parts["live_playback"]{
          if let hlsvp = parts["hlsvp"] as? String {
            return [
                "url": "\(hlsvp)" as AnyObject,
                "title": "\(videoTitle)" as AnyObject,
                "image": "\(streamImage)" as AnyObject,
                "isStream": true as AnyObject
            ]
          }
        } else {
            let fmtStreamMapArray = fmtStreamMap.components(separatedBy: ",")
          for videoEncodedString in fmtStreamMapArray {
            var videoComponents = videoEncodedString.dictionaryFromQueryStringComponents()
            videoComponents["title"] = videoTitle as AnyObject
            videoComponents["isStream"] = false as AnyObject
            return videoComponents as [String: AnyObject]
          }
        }
      }
    }
    return nil
  }
  
  /**
  Block based method for retreiving a iOS supported video link

  @param youtubeURL the the complete youtube video url
  @param completeBlock the block which is called on completion

  */
  public static func h264videosWithYoutubeURL(youtubeURL: NSURL,completion: ((
    _ videoInfo: [String: AnyObject]?, _ error: NSError?) -> Void)?) {
    
    DispatchQueue.global(qos: .background).async {
        if let youtubeID = self.youtubeIDFromYoutubeURL(youtubeURL: youtubeURL), let videoInformation = self.h264videosWithYoutubeID(youtubeID: youtubeID) {
            DispatchQueue.main.async {
                completion?(videoInformation, nil)
            }
        }else{
            DispatchQueue.main.async {
                 completion?(nil, NSError(domain: "com.player.youtube.backgroundqueue", code: 1001, userInfo: ["error": "Invalid YouTube URL"]))
            }
        }
    }
    /*
    let priority = DispatchQueue.GlobalQueuePriority.background
    dispatch_get_global_queue(priority, 0).async() {
        if let youtubeID = self.youtubeIDFromYoutubeURL(youtubeURL: youtubeURL), let videoInformation = self.h264videosWithYoutubeID(youtubeID) {
            dispatch_get_main_queue().async() {
            completion?(videoInfo: videoInformation, error: nil)
          }
        }else{
            dispatch_get_main_queue().async() {
            completion?(nil, NSError(domain: "com.player.youtube.backgroundqueue", code: 1001, userInfo: ["error": "Invalid YouTube URL"]))
          }
        }
      }
 */
    }
  }


YoutubeSourceParserKitTests/YoutubeSourceParserKitTests.swift
//
//  youtube_parserTests.swift
//  youtube-parserTests
//
//  Created by Toygar Dündaralp on 7/5/15.
//  Copyright (c) 2015 Toygar Dündaralp. All rights reserved.
//

import UIKit
import XCTest

class YoutubeSourceParserKitTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
    }
    
    override func tearDown() {
        super.tearDown()
    }

  func testStringByDecodingURLFormat() {
    let originalString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ+XUFpdwbfqqQ"
    let testString = "https://www.youtube.com/watch?v=XUFpdwbfqqQ XUFpdwbfqqQ"
    let comparsionString = originalString.stringByDecodingURLFormat()
    XCTAssertNotNil(comparsionString, "comparsionString is nil")
    XCTAssertEqual(comparsionString, testString, "String decoding failed")
  }

  func testDictionaryFromQueryStringComponents() {
    let sampleLink = "https://www.youtube.com/watch?v=o0jJiB2Ygpg&list=RDo0jJiB2Ygpg"
    let dictionary = sampleLink.dictionaryFromQueryStringComponents() as [String:AnyObject]
    XCTAssertNotNil(dictionary["list"], "url dictionary parse error")
    if let list = dictionary["list"] as? String {
      XCTAssertEqual(list, "RDo0jJiB2Ygpg", "list not equal value")
    }
    if let link = dictionary["https://www.youtube.com/watch?v="] as? String {
      XCTAssertEqual(link, "o0jJiB2Ygpg", "link not equal value")
    }
  }

  func testInvalidURLs1() {
    let sampleLink = NSURL(string: "?v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs2() {
    let sampleLink = NSURL(string: "?v=")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "", "Youtube ID not same")
    }
  }

  func testInvalidURLs3() {
    let sampleLink = NSURL(string: "v=1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "v=1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testInvalidURLs4() {
    let sampleLink = NSURL(string: "v1hZ98an9wjo")!
    XCTAssertNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is not nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "v1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromYoutubeURL() {
    let sampleLink = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testYoutubeIDFromMobileYoutubeURL() {
    let sampleLink = NSURL(string: "https://m.youtube.com/#/watch?v=1hZ98an9wjo")!
    XCTAssertNotNil(Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink), "Youtube ID is nil")
    if let youtubeID = Youtube.youtubeIDFromYoutubeURL(youtubeURL: sampleLink) {
      XCTAssertEqual(youtubeID, "1hZ98an9wjo", "Youtube ID not same")
    }
  }

  func testH264videosWithYoutubeURL() {
    if let videoComponents = Youtube.h264videosWithYoutubeID(youtubeID: "1hZ98an9wjo") {
      XCTAssertNotNil(videoComponents, "video component is nil")
      if let itag = videoComponents["itag"] as? String {
        XCTAssertEqual(itag, "22", "itag not equal")
      }
      if let quality = videoComponents["quality"] as? String {
        XCTAssertEqual(quality, "hd720", "quality not equal")
      }
      if let fallback_host = videoComponents["fallback_host"] as? String {
        XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
      }
      if let type = videoComponents["type"] as? String {
        XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
      }
    }
  }

  func testVideoQuality(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
        Youtube.h264videosWithYoutubeURL(youtubeURL: videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["quality"] as? String {
            XCTAssertEqual(quality, "hd720", "quality not equal")
          }
        }
      })
    }
  }

  func testItagControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
        Youtube.h264videosWithYoutubeURL(youtubeURL: videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let itag = info["itag"] as? String {
            XCTAssertEqual(itag, "22", "itag not equal")
          }
        }
      })
    }
  }

  func testVideoTypeControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
        Youtube.h264videosWithYoutubeURL(youtubeURL: videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let type = info["type"] as? String {
            XCTAssertEqual(type, "video/mp4; codecs=\"avc1.64001F, mp4a.40.2\"", "type not equal")
          }
        }
      })
    }
  }

  func testFallbackHostControl(){
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
        Youtube.h264videosWithYoutubeURL(youtubeURL: videoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          if let fallback_host = info["fallback_host"] as? String {
            XCTAssertEqual(fallback_host, "tc.v7.cache1.googlevideo.com", "fallback_host not equal")
          }
        }
      })
    }
  }

  func testIsLiveVideoTest() {
    if let liveVideoURL = NSURL(string: "https://www.youtube.com/watch?v=rxGoGg7n77A"){
        Youtube.h264videosWithYoutubeURL(youtubeURL: liveVideoURL, completion: { (videoInfo, error) -> Void in
        XCTAssertNotNil(videoInfo, "video dictionary is nil")
        if let info = videoInfo as [String:AnyObject]? {
          XCTAssertNotNil(info["url"], "live stream url is nil")
        }
      })
    }
  }

  func testInfoUrl(){
    XCTAssertEqual(Youtube.infoURL, "http://www.youtube.com/get_video_info?video_id=", "Info url not correct!")
  }

  func testUserAgent() {
     XCTAssertEqual(Youtube.userAgent, "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_2) AppleWebKit/537.4 (KHTML, like Gecko) Chrome/22.0.1229.79 Safari/537.4", "User Agent not correct!")
  }

  func testH264videosWithYoutubeURLBlock(){
    let expectation: XCTestExpectation = self.expectation(description: "Handler called")
    if let videoURL = NSURL(string: "http://www.youtube.com/watch?v=1hZ98an9wjo") {
        Youtube.h264videosWithYoutubeURL(youtubeURL: videoURL, completion: { (videoInfo, error) -> Void in
        expectation.fulfill()
        if let info = videoInfo as [String:AnyObject]? {
          if let quality = info["itag"] as? String {
            XCTAssertEqual(quality, "22", "itag not equal")
          }
        }
      })
        self.waitForExpectations(timeout: 2, handler: nil)
    }

  }
  
}
